### ISSUE TITLE
Add ability to recover from errors

### ISSUE THREAD

**Issue by nrzull** [telegraf/telegraf#1657]
**Is your feature request related to a problem? Please describe.**
For now, if your bot receives, for example, TelegramError, it will crash your entire app and if you have multiple bots then you will not even know which bot crashed (with `process.on("unhandledRejection", (e) => {})`). Related code snippet:

```javascript
        // src/core/network/polling.ts
        if (
          err instanceof TelegramError &&
          // Unauthorized      Conflict
          (err.code === 401 || err.code === 409)
        ) {
          this.skipOffsetSync = true
          throw err
        } else {
          throw err
        }
```

**Describe the solution you'd like**
Would be cool to add ability to handle such errors. It also may have restart method

```javascript
const bot = new Telegraf(token);

bot.on("unhandledRejectionOrAnyOtherReasonThatKilledOurBot", (error) => {
 // here bot is dead, so we could restart him
 bot.restart()
} )
```

**Describe alternatives you've considered**
-

**Additional context**

**Comment by hypernova7** [comment]
You can just do this:

```javascript
const launchOptions = {
  webhook: {
    // ...
  }
}
bot.launch(launchOptions)

process.on('unhandledRejectionOrAnyOtherReasonThatKilledOurBot', (error) => {
  bot.launch(launchOptions)
})
```
You can also make use of `bot.catch()` to prevent your bot from stopping due to an error, just try to use `async/await`

```javascript
bot.catch(error => console.error(error))
bot.help(async ctx => {
  await ctx.reply('Hi')
})
```

**Comment by nrzull** [comment]
@hypernova7 Well, `catch` method will not catch `TelegramError` that crashes your app. At least it didn't help me  :shrug:

Your first snippet is a "fine" workaround, but if I have pool of bots then I will not know which one was stopped. I could register one `unhandledRejection` per one bot creation, but for me such solution looks "smelly".

Anyway, thanks for reply!

**Comment by MKRhere** [comment]
`bot.catch()` catches `TelegramError`, but you must always `await` all your promises for it to work. For example, `await ctx.reply(...)` or `await next()`. If you don't add await/return for all async calls, Telegraf is unable to handle the error, and only Node can. That's why it's an `unhandledRejection`. Telegraf cannot catch `unhandledRejection` per bot, since it is per-process.

**Comment by AshenCoon** [comment]
`bot.catch()`  will not catch this error, it will always be unhandled. It is thrown here
`src/core/network/polling.ts#L77`
```node
async loop(handleUpdate: (updates: tg.Update) => Promise<void>) {
    if (this.abortController.signal.aborted) {
      throw new Error('Polling instances must not be reused!')
    }
    try {
      for await (const updates of this) {
        await Promise.all(updates.map(handleUpdate))
      }
    } finally {
      debug('Long polling stopped')
      // prevent instance reuse
      this.stop()
      await this.syncUpdateOffset().catch(noop)
    }
  }
```
and nothing is set to handle it here
`src/telegraf.ts#L208`
```node
this.polling.loop(async (update) => {
      await this.handleUpdate(update)
    })
```
only errors in handleUpdate will be handled, and not anything thrown in polling. Ofcourse you can catch app-wide unhandled errors and restart it, but it is really problematic if you have a lot of bots in one app. And its looks wrong anyway.

**Comment by wojpawlik** [comment]
How exactly do you plan to handle these errors? They're fatal for a reason.

**Comment by MKRhere** [comment]
This was released to stable as [v4.11.0](https://github.com/telegraf/telegraf/releases/tag/v4.11.0), feel free to update now :)

**Comment by aSKz1** [comment]
> **Is your feature request related to a problem? Please describe.**
> For now, if your bot receives, for example, TelegramError, it will crash your entire app and if you have multiple bots then you will not even know which bot crashed (with `process.on("unhandledRejection", (e) => {})`). Related code snippet:
> 
> ```javascript
>         // src/core/network/polling.ts
>         if (
>           err instanceof TelegramError &&
>           // Unauthorized      Conflict
>           (err.code === 401 || err.code === 409)
>         ) {
>           this.skipOffsetSync = true
>           throw err
>         } else {
>           throw err
>         }
> ```
> 
> **Describe the solution you'd like**
> Would be cool to add ability to handle such errors. It also may have restart method
> 
> ```javascript
> const bot = new Telegraf(token);
> 
> bot.on("unhandledRejectionOrAnyOtherReasonThatKilledOurBot", (error) => {
>  // here bot is dead, so we could restart him
>  bot.restart()
> } )
> ```
> 
> **Describe alternatives you've considered**
> -
> 
> **Additional context**
> 
>
